### DISCUSSION TITLE
Understanding worker initialization and connection

### DISCUSSION THREAD

**Discussion by wlandau** [mschubert/clustermq#290] [Q&A]
From https://github.com/mschubert/clustermq/issues/208#issuecomment-725444690 and https://mschubert.github.io/clustermq/articles/technicaldocs.html#worker-startup, I understand this part of the worker initialization process:

1. The scheduler launches a new job.
2. The job launches R and calls `clustermq:::worker("IP_OF_LOCAL_PROCESS")` to connect back to the local R process.

After that, how exactly does the tunneling and reverse tunneling work? I am not sure I understand https://github.com/mrc-ide/rrq/issues/64#issuecomment-1284411299:

> In clustermq the tunnel is established local>remote, so no need to expose any local ports (but the remote needs to accept SSH connections and reverse tunneling; traffic over the tunnel is always encrypted via SSH)

For the tunneling that happens from the local machine to the remote machine, is that a step after (2)? I thought (2) involved a new SSH connection from the remote process to the local process.

**Comment by mschubert** [comment]
There are two options, either the workers are on the same network (then they need to know the host ip/port to connect to), or they are on a separate network (then they need to know the ip/port *of the remote end of the tunnel*).

The tunneling is "step 0", so it occurs before launching workers. It is active before because it launches the proxy, which in turn launches the workers. The tunnel and proxy will stay active because they forward worker traffic to/from the host (using two reverse tunnels, one for controlling the proxy and one for forwarding traffic).

This is why, in the *SSH.tmpl* file, you can see:

https://github.com/mschubert/clustermq/blob/78125771cda27373bf19eb0f38ac5106524d4eae/inst/SSH.tmpl#L2-L3

Workers behind a proxy (tunnel) need to connect to the `fwd_port` of the tunnel *on the remote end*. So there is no need to expose any ports from the user to the remote.

**Reply by wlandau** [reply]
I see now, https://github.com/mrc-ide/rrq/issues/64#issuecomment-1284411299 describes the SSH connector.

If I understand correctly, `ctl_port` on the proxy forwards data to `local_port` on the local machine (to control the proxy), and `job_port` on the proxy is forwarded to `fwd_port` on the on the local machine (to manage jobs). Are these tunnels bidirectional? I'm wondering if data can travel from `localhost:local_port` to `proxy_ip:ctl_port` *AND* from `proxy_ip:ctl_port` to `localhost:local_port`. If so, then what is the difference between forward tunneling and reverse tunneling? Is it just a matter of who initiates the connection and which IP address needs to be known in advance?

**Reply by mschubert** [reply]
Yes exactly, the directionality comes from which end can establish a TCP/IP connection via the tunnel. Once this is established, both peers can send/receive data.

**Reply by wlandau** [reply]
Got it, thanks for confirming.

I think I understand why reverse tunnels are needed when it comes to proxy-worker communication: the worker knows the IP address of the proxy, but not the other way around. But what about the local-to-proxy connection? From https://mschubert.github.io/clustermq/articles/userguide.html#ssh-connector, it looks like `clustermq.ssh.host` specifies one of the peers. If `clustermq.ssh.host` is the local machine, then how does the local machine know where to find the proxy? If `clustermq.ssh.host` is the proxy, then why is a reverse tunnel necessary, as opposed to a forward tunnel?

**Reply by mschubert** [reply]
`clustermq.ssh.host` is the host to establish the SSH connection to, as defined in your *~/.ssh/config* (not any of the peers). It is the same host the proxy will be running on, and the remote tunnel ends will be opened.

We need a reverse tunnel because we still won't know the workers' IPs, so they need to connect to the remote tunnel end (= reverse tunneling).

**Reply by wlandau** [reply]
Thanks, Michael.

I'm afraid I'm still a bit confused. The proxy-to-local tunnels are both created prior to all the worker-to-proxy tunnels, right? So wouldn't the worker be able to know about `fwd_port` regardless of whether `-R {{ job_port }}:localhost:{{ fwd_port }` is a forward or reverse tunnel?

Sorry if I am missing something basic about how SSH works. Or maybe an architecture diagram in the technical documentation would help me understand.

**Reply by wlandau** [reply]
Still wrapping my head around which ports are on which computers and which computers initiate the tunnels.

**Reply by mschubert** [reply]
I think you're misunderstanding the difference between a forward (`-L`) and a reverse tunnel (`-R`). In both cases, the SSH connection is established local>remote. With a forward tunnel, I can locally connect to a port that will be forwarded to the remote. With a reverse tunnel, a remote process can connect to a remote port that will be forwarded to the local machine.

Does that help?

**Reply by wlandau** [reply]
It does help, thanks. So the tunnel that forwards traffic needs to be a reverse tunnel because the workers need to find a port on the proxy that will forward data from the proxy to the local machine (fwd_port?).

**Reply by wlandau** [reply]
Then in what sense are connections bidirectional?

**Reply by mschubert** [reply]
Once the connection is established, both peers can send/receive data (with some limitations imposed by ZMTP)
